api/envoy/extensions/filters/http/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,80 @@
+
+import "envoy/config/core/v3/config_source.proto";
+import "envoy/config/core/v3/grpc_service.proto";
+import "envoy/extensions/filters/http/mutation/v3alpha/processing_mode.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.mutation.v3alpha";
+option java_outer_classname = "MutationProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).work_in_progress = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Mutation Service]
+// Mutation Service
+// [#extension: envoy.filters.http.mutation]
+
+// The Mutation filter allows an external service to act on HTTP traffic in a very flexible way.
+// It communicates with an external gRPC service that can use it to do a variety of things
+// with the request and response:
+//
+// * Access and modify the HTTP headers on the request, response, or both
+// * Access and modify the HTTP request and response bodies
+// * Access and modify the dynamic stream metadata
+// * Immediately send an HTTP response downstream and terminate other processing
+//
+// The filter communicates with the server using a gRPC bidirectional stream. After the initial
+// request, the external server is in control over what additional data is sent to it
+// and how it should be processed.
+//
+// By implementing the protocol specified by the stream, the external server can choose:
+//
+// * Whether it receives the response message at all
+// * Whether it receives the message body at all, in separate chunks, or as a single buffer
+// * Whether subsequent HTTP requests are transmitted synchronously or whether they are
+//   sent asynchronously.
+//
+// All of this together allows a server to process the filter traffic in fairly
+// sophisticated ways. For example:
+//
+// * A server may choose to examine all or part of the HTTP message bodies depending
+//   on the content of the headers.
+// * A server may choose to immediately reject some messages based on their HTTP
+//   headers (or other dynamic metadata) and more carefully examine others
+// * A server may asynchronously monitor traffic coming through the filter by inspecting
+//   headers, bodies, or both, and then decide to switch to a synchronous processing
+//   mode, either permanently or temporarily.
+
+// [#next-free-field: 7]
+message Mutation {
+  // Configuration for the gRPC service that the filter will communicate with.
+  // The filter supports both the "Envoy" and "Google" gRPC clients.
+  config.core.v3.GrpcService grpc_service = 1;
+
+  // The version of the API supported by the gRPC service.
+  config.core.v3.ApiVersion transport_api_version = 2
+      [(validate.rules).enum = {defined_only: true}];
+
+  // If true, then do nothing if the remote service RPC fails.
+  // Otherwise, return a 500 error if the gRPC call fails.
+  bool failure_mode_allow = 3;
```

Comment:
this would also call into question the name of the filter, but would it make sense to support a "read only" mode? then this filter could be used for logging and not allowed to actually mutate anything? This could also be implemented with `failure_mode_allow` being true and the service always returning the `CONTINUE_*` responses, but it might be nice to have a guarantee against the upstream modifying anything in certain cases?

Reply:
@gbrail I missed the most recent updates, is this now ready for a full review again? I can take a look tomorrow.

I think we really should change the name. As @mpuncel points out, it won't always mutate. Also, it makes it hard to discover; we had a question at EnvoyCon today about whether there is a "gRPC service for offloading compute" as an alternative to Wasm, I shared this PR. I vote for "gRPC remote filter" (grpc_remote) or "remote filter" (remote), but there are possibly better names.

Reply:
+1 commented on this above.

Reply:
"external processing filter," or "ext_proc" is the leading candidate.